### ISSUE TITLE
TOCO converter requires min/max information for taking output from intermediate nodes?

### ISSUE THREAD

**Issue by liquidscience** [tensorflow/tensorflow#18979]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:No.
- **TensorFlow version (use command below)**:v1.7.0-3-g024aecf414 1.7.0
- **Python version**: 3.5.4
- **Bazel version (if compiling from source)**:N/A
- **GCC/Compiler version (if compiling from source)**:N/A
- **CUDA/cuDNN version**:9.0
- **GPU model and memory**:Nvidia 840M
- **Exact command to reproduce**:N/A

### Describe the problem
I have a quantized mobilenet downloaded from the official repository([here](https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet_v1.md)) , I want to collect output from the last pointwise convolutional layer of the network and run on a mobile.

The quantized frozen model contains additional fc,softmax etc layers that are of no use for my application.

To convert my quantized and frozen graph to specified input & output i run the below command
```
toco \
  --input_file=mobilenet_v1_1.0_224_quant_frozen.pb \
  --output_file=corrected_mobilenet_v1_1.0_224_frozen.tflite \
  --input_format=TENSORFLOW_GRAPHDEF --output_format=TFLITE \
  --inference_type=QUANTIZED_UINT8 \
  --input_shape="1,224,224,3" \
  --input_array=input \
  --output_array=MobilenetV1/MobilenetV1/Conv2d_13_pointwise/Conv2D_Fold \
  --std_value=127.5 --mean_value=127.5
```
and receive the following output from the converter
```
WARNING:tensorflow:From /home/sam/.virtualenvs/tf6/lib/python3.5/site-packages/tensorflow/contrib/learn/python/learn/datasets/base.py:198: retry (from tensorflow.contrib.learn.python.learn.datasets.base) is deprecated and will be removed in a future version.
Instructions for updating:
Use the retry module or similar alternatives.
2018-04-30 16:31:24.735099: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 598 operators, 889 arrays (0 quantized)
2018-04-30 16:31:24.753483: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After Removing unused ops pass 1: 571 operators, 851 arrays (0 quantized)
2018-04-30 16:31:24.773414: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 571 operators, 851 arrays (0 quantized)
2018-04-30 16:31:25.122835: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 54 operators, 109 arrays (1 quantized)
2018-04-30 16:31:25.123855: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before pre-quantization graph transformations: 54 operators, 109 arrays (1 quantized)
2018-04-30 16:31:25.124206: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After pre-quantization graph transformations pass 1: 28 operators, 83 arrays (1 quantized)
2018-04-30 16:31:25.124569: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before quantization graph transformations: 28 operators, 83 arrays (1 quantized)
2018-04-30 16:31:25.155063: F tensorflow/contrib/lite/toco/graph_transformations/quantize.cc:169] Array MobilenetV1/MobilenetV1/Conv2d_13_pointwise/Conv2D_Fold does not have MinMax information, and is not a constant array. Cannot proceed with quantization.
Aborted (core dumped)
```

Why does this quantized graph requires min max information for intermediate nodes when it can resolve the final nodes correctly without additional min/max information?

I also tried manually giving default min/max information but i suffer a huge performance drop if i do this.

**Comment by bjacob** [comment]
The flag,

```--output_array=MobilenetV1/MobilenetV1/Conv2d_13_pointwise/Conv2D_Fold```

is asking toco to cut this graph at this point, i.e. immediately after the Conv2D op, before the subsequent BiasAdd, ReLU6... and FakeQuantWithMinMaxVars.

The resulting cropped graph thus does not have anymore the required min-max info for the output node.

Passing arbitrary --input_arrays , --output_arrays to cut graphs is fine, as long as the required granularity of the graph is preserved. When quantizing (--inference_type=QUANTIZED_UINT8), that required granularity is whichever granularity FakeQuant nodes are inserted at, e.g. typically whole fused layers (Conv2D+BiasAdd+Relu6).

Try this instead:

```
--output_array=MobilenetV1/MobilenetV1/Conv2d_13_pointwise/act_quant/FakeQuantWithMinMaxVars
```

**Comment by bjacob** [comment]
There is no 1:1 mapping of TensorFlow nodes to quantized operations.
TensorFlow nodes are fused in a specific way (Conv+Bias+Relu) that corresponds to the placement of FakeQuant nodes; this is what allows the training process to effectively train for quantization.
Thus, when performing quantized inference, it is not possible to let --output_array point to arbitrary nodes in the original TF graph, as these nodes just don't exist in the quantized inference workload.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
Closing due to inactivity. Please re-open if there are remaining questions we have missed. Thanks!
